src/content/docs/develop/calling-frontend.mdx

```diff
@@ -2,11 +2,404 @@
+  download_id: usize,
+  content_length: usize,
+}
+
+#[derive(Clone, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct DownloadProgress {
+  download_id: usize,
+  chunk_length: usize,
+}
+
+#[derive(Clone, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct DownloadFinished {
+  download_id: usize,
+}
+
+#[tauri::command]
+fn download(app: AppHandle, url: String) {
+  let content_length = 1000;
+  let download_id = 1;
+
+  app.emit("download-started", DownloadStarted {
+    url: &url,
+    download_id,
+    content_length
+  }).unwrap();
+
+  for chunk_length in [15, 150, 35, 500, 300] {
+    app.emit("download-progress", DownloadProgress {
+      download_id,
+      chunk_length,
+    }).unwrap();
+  }
+
+  app.emit("download-finished", DownloadFinished { download_id }).unwrap();
+}
+```
+
+### Listening to Events
+
+Tauri provides APIs to listen to events on both the webview and the Rust interfaces.
+
+#### Listening to Events on the Frontend
+
+The `@tauri-apps/api` NPM package offers APIs to listen to both global and webview-specific events.
+
+- Listening to global events
+
+  ```ts
+  import { listen } from '@tauri-apps/api/event';
+
+  type DownloadStarted = {
+    url: string;
+    downloadId: number;
+    contentLength: number;
+  };
+
+  listen<DownloadStarted>('download-started', (event) => {
+    console.log(
+      `downloading ${event.payload.contentLength} bytes from ${event.payload.url}`
+    );
+  });
+  ```
+
+- Listening to webview-specific events
+
+  ```ts
+  import { getCurrentWebviewWindow } from '@tauri-apps/api/webviewWindow';
+
+  const appWebview = getCurrentWebviewWindow();
+  appWebview.listen<string>('logged-in', (event) => {
+    localStorage.setItem('session-token', event.payload);
+  });
+  ```
+
+The `listen` function keeps the event listener registered for the entire lifetime of the application.
+To stop listening on an event you can use the `unlisten` function which is returned by the `listen` function:
+
+```js
+import { listen } from '@tauri-apps/api/event';
+
+const unlisten = await listen('download-started', (event) => {});
+unlisten();
+```
+
+:::note
+Always use the unlisten function when your execution context goes out of scope
+such as when a component is unmounted.
+
+When the page is reloaded or you navigate to another URL the listeners are unregistered automatically.
+This does not apply to a Single Page Application (SPA) router though.
+:::
+
+Additionally Tauri provides a utility function for listening to an event exactly once:
+
+```js
+import { once } from '@tauri-apps/api/event';
+import { getCurrentWebviewWindow } from '@tauri-apps/api/webviewWindow';
+
+once('ready', (event) => {});
+
+const appWebview = getCurrentWebviewWindow();
+appWebview.once('ready', () => {});
+```
+
+:::note
+Events emitted in the frontend also triggers listeners registed by these APIs.
+For more information see the [Calling Rust from the Frontend] documentation.
+:::
+
+#### Listening to Events on Rust
```

Comment:
Someone wanna bet on how long it'll take for someone to interpret this as instructions for a Rust frontend? :D

Reply:
indeed we should enhance the wording here